src/lrucache.h

```diff
@@ -70,51 +79,71 @@ class lru_cache {
   };
 
 public: // functions
-	explicit lru_cache(size_t max_size) :
-		_max_size(max_size) {
-	}
-
-	void put(const key_t& key, const value_t& value) {
-		auto it = _cache_items_map.find(key);
-		if (it != _cache_items_map.end()) {
-			_cache_items_list.splice(_cache_items_list.begin(), _cache_items_list, it->second);
+  explicit lru_cache(size_t max_size) :
+    _max_size(max_size) {
+  }
+
+  // If 'key' is present in the cache, returns the associated value,
+  // otherwise puts the given value into the cache (and returns it with
+  // a status of a cache miss).
+  AccessResult getOrPut(const key_t& key, const value_t& value) {
+    auto it = _cache_items_map.find(key);
+    if (it != _cache_items_map.end()) {
+      _cache_items_list.splice(_cache_items_list.begin(), _cache_items_list, it->second);
+      return AccessResult(it->second->second, HIT);
+    } else {
+      putMissing(key, value);
+      return AccessResult(value, PUT);
+    }
+  }
+
+  void put(const key_t& key, const value_t& value) {
+    auto it = _cache_items_map.find(key);
+    if (it != _cache_items_map.end()) {
+      _cache_items_list.splice(_cache_items_list.begin(), _cache_items_list, it->second);
       it->second->second = value;
-		} else {
-      _cache_items_list.push_front(key_value_pair_t(key, value));
-		  _cache_items_map[key] = _cache_items_list.begin();
-      if (_cache_items_map.size() > _max_size) {
-        auto last = _cache_items_list.end();
-        last--;
-        _cache_items_map.erase(last->first);
-        _cache_items_list.pop_back();
-      }
+    } else {
+      putMissing(key, value);
+    }
+  }
+
+  AccessResult get(const key_t& key) {
+    auto it = _cache_items_map.find(key);
+    if (it == _cache_items_map.end()) {
+      return AccessResult();
+    } else {
+      _cache_items_list.splice(_cache_items_list.begin(), _cache_items_list, it->second);
+      return AccessResult(it->second->second, HIT);
+    }
+  }
+
+  bool exists(const key_t& key) const {
+    return _cache_items_map.find(key) != _cache_items_map.end();
+  }
+
+  size_t size() const {
+    return _cache_items_map.size();
+  }
+
+private: // functions
+  void putMissing(const key_t& key, const value_t& value) {
+    assert(_cache_items_map.find(key) == _cache_items_map.end());
+    _cache_items_list.push_front(key_value_pair_t(key, value));
+    _cache_items_map[key] = _cache_items_list.begin();
+    if (_cache_items_map.size() > _max_size) {
+      auto last = _cache_items_list.end();
+      last--;
```

Comment:
Can we do `auto last = _cache_items_list.back()` instead ?

Reply:
Agree